crates/execution/src/matching_engine/engine.rs

```diff
@@ -1106,45 +1106,129 @@ impl OrderMatchingEngine {
+                        }
+                        return true;
+                    }
+                    return false;
+                }
+
+                let ap = inner.activation_price.unwrap();
+                let hit = match inner.order_side() {
+                    OrderSide::Buy => ask.is_some_and(|a| a <= ap),
+                    OrderSide::Sell => bid.is_some_and(|b| b >= ap),
+                    _ => false,
+                };
+                if hit {
+                    inner.set_activated();
+                    if let Err(e) = self.cache.borrow_mut().update_order(order) {
+                        log::error!("Failed to update order: {e}");
+                    }
+                }
+                hit
+            }
+            _ => true,
+        }
+    }
+
     fn iterate_orders(&mut self, timestamp_ns: UnixNanos, orders: &[PassiveOrderAny]) {
         for order in orders {
             if order.is_closed() {
                 continue;
             }
 
-            // Check expiration
-            if self.config.support_gtd_orders
-                && let Some(expire_time) = order.expire_time()
-                && timestamp_ns >= expire_time
-            {
-                // SAFTEY: We know this order is in the core
-                self.core.delete_order(order).unwrap();
-                self.cached_filled_qty.remove(&order.client_order_id());
-                self.expire_order(order);
+            if self.config.support_gtd_orders {
+                if let Some(expire_time) = order.expire_time() {
+                    if timestamp_ns >= expire_time {
+                        let _ = self.core.delete_order(order);
+                        self.cached_filled_qty.remove(&order.client_order_id());
+                        self.expire_order(order);
+                        continue;
+                    }
+                }
             }
 
-            // Manage trailing stop
-            if let PassiveOrderAny::Stop(o) = order
-                && let PassiveOrderAny::Stop(
-                    StopOrderAny::TrailingStopMarket(_) | StopOrderAny::TrailingStopLimit(_),
-                ) = order
-            {
-                let mut order = OrderAny::from(o.to_owned());
-                self.update_trailing_stop_order(&mut order);
+            if matches!(
+                order,
+                PassiveOrderAny::Stop(
+                    StopOrderAny::TrailingStopMarket(_) | StopOrderAny::TrailingStopLimit(_)
+                )
+            ) {
+                let mut any = OrderAny::from(order.clone());
+
+                if !self.maybe_activate_trailing_stop(&mut any, self.core.bid, self.core.ask) {
+                    continue;
+                }
+
+                self.update_trailing_stop_order(&mut any);
             }
 
-            // Move market back to targets
-            if let Some(target_bid) = self.target_bid {
-                self.core.bid = Some(target_bid);
-                self.target_bid = None;
```

Comment:
There could be a reason for explicitly clearing this, but it'll be picked up when serious focus is put back onto the execution side.